File: binary_tree/merge_two_binary_trees/solution2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func mergeTrees(root1 *TreeNode, root2 *TreeNode) *TreeNode {
    var dfs func(node1, node2 *TreeNode)
	dfs = func(node1, node2 *TreeNode) {
		if node2 == nil {
			return
		}

		node1.Val += node2.Val
		if node2.Left != nil && node1.Left == nil {
			node1.Left = &TreeNode{}
		}
		if node2.Right != nil && node1.Right == nil {
			node1.Right = &TreeNode{}
		}
		dfs(node1.Left, node2.Left)
		dfs(node1.Right, node2.Right)
	}
	if root1 == nil {
		root1, root2 = root2, root1
	}

	dfs(root1, root2)
	return root1
}

/*
	执行用时分布：0ms，击败 100.00%
	消耗内存分布：8.38MB，击败 87.57%
*/

File: binary_tree/insufficient_nodes_in_root_to_leaf_paths/solution2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
min_val = -10**7

class Solution:
    def sufficientSubset(self, root: Optional[TreeNode], limit: int) -> Optional[TreeNode]:
        self.traverse(root, 0, limit)
        if root.val == min_val:
            return None
        self.remove(root)

        return root
        
    def traverse(self, root: Optional[TreeNode], parent_sum: int, limit: int) -> int:
        cur_sum = min_val
        if root.left is None and root.right is None:
            cur_sum = parent_sum+root.val
        else:
            if root.left is not None:
                cur_sum = self.traverse(root.left, parent_sum+root.val, limit) if root.left is not None else min_val
            if root.right is not None:
                cur_sum = max(cur_sum, self.traverse(root.right, parent_sum+root.val, limit) if root.right is not None else min_val)    

        if cur_sum < limit:
            root.val = min_val
        
        return cur_sum

    def remove(self, root: Optional[TreeNode]):
        if root is None:
            return

        if root.left is not None and root.left.val == min_val:
            root.left = None
        elif root.right is not None and root.right.val == min_val:
            root.right = None

        self.remove(root.left)
        self.remove(root.right)    

# 执行用时分布：4ms 击败 55.41%
# 消耗内存分布：20.27MB，击败 25.54%